rdt-app/src/main/java/io/ona/rdt_app/widget/RDTExpirationDateReaderFactory.java

```diff
@@ -0,0 +1,160 @@
+        String key = jsonObject.getString(JsonFormConstants.KEY);
+
+        rootLayout.setTag(com.vijay.jsonwizard.R.id.key, key);
+        rootLayout.setTag(com.vijay.jsonwizard.R.id.openmrs_entity_parent, openMrsEntityParent);
+        rootLayout.setTag(com.vijay.jsonwizard.R.id.openmrs_entity, openMrsEntity);
+        rootLayout.setTag(com.vijay.jsonwizard.R.id.openmrs_entity_id, openMrsEntityId);
+    }
+
+
+    private OnActivityResultListener createOnActivityResultListener() {
+
+        OnActivityResultListener resultListener =  new OnActivityResultListener() {
+
+            @Override
+            public void onActivityResult(int requestCode, int resultCode, Intent data) {
+                hideProgressDialog();
+                final JsonApi jsonApi = (JsonApi) widgetArgs.getContext();
+                if (requestCode == JsonFormConstants.RDT_CAPTURE_CODE && resultCode == RESULT_OK) {
+                    if (data != null) {
+                        try {
+                            String imageFilePath = data.getExtras().getString(SAVED_IMAGE_FILE_PATH);
+                            widgetArgs.getJsonObject().put(VALUE, imageFilePath);
+                            // write image path as widget value
+                            String key = (String) rootLayout.getTag(com.vijay.jsonwizard.R.id.key);
+                            String openMrsEntityParent = (String) rootLayout.getTag(com.vijay.jsonwizard.R.id.openmrs_entity_parent);
+                            String openMrsEntity = (String) rootLayout.getTag(com.vijay.jsonwizard.R.id.openmrs_entity);
+                            String openMrsEntityId = (String) rootLayout.getTag(com.vijay.jsonwizard.R.id.openmrs_entity_id);
+                            jsonApi.writeValue(widgetArgs.getStepName(), key, imageFilePath, openMrsEntityParent,
+                                    openMrsEntity, openMrsEntityId, widgetArgs.isPopup());
+                        } catch (JSONException e) {
+                            Log.e(TAG, e.getStackTrace().toString());
+                        }
+                    } else {
+                        Log.i(TAG, "No result data for RDT capture!");
+                    }
+                    if (!widgetArgs.getFormFragment().next()) {
+                        widgetArgs.getFormFragment().save(true);
+                    }
+                } else if (resultCode == RESULT_CANCELED) {
+                    ((Activity) widgetArgs.getContext()).finish();
+                }
+            }
+        };
+
+        return resultListener;
+    }
+
+    public void setUpRDTExpirationDateActivity() {
+        Context context = widgetArgs.getContext();
+        if (context instanceof JsonApi) {
+            final JsonApi jsonApi = (JsonApi) context;
+            jsonApi.addOnActivityResultListener(JsonFormConstants.RDT_CAPTURE_CODE , createOnActivityResultListener());
+        }
+    }
+
+    private class LaunchRDTCameraTask extends AsyncTask<Void, Void, Void> {
```

Comment:
Why start the RDTExpirationDateActivity in a background thread?

Reply:
The camera takes time to launch.

This provides a way to show a progress dialog as the activity waits to launch.